### PR TITLE
DeviceSourceManager: Fix discrepancy with onInputChangedObservable contract

### PR DESCRIPTION
**Pull request by PolygonalSun**
With the changes made for 5.0, there was a change made to some parts of the DeviceSourceManager that changed the contract for onInputChangedObservable.  This PR contains changes to fix that discrepancy and remove IDeviceEvent and INativeInput.

This PR also changes the DeviceSourceManager to keep track of its sources, rather than referencing ones on the InternalDeviceSourceManager.  This should prevent observers from staying when a specific DeviceSourceManager instance is disposed of.

### CODE REVIEW

File: guiEditor/src/diagram/workbench.tsx [hunk 1]
```diff
@@ -853,8 +854,16 @@ export class WorkbenchComponent extends React.Component<IWorkbenchComponentProps
     //Add zoom and pan controls
     addControls(scene: Scene) {
 
-        const zoomFnScrollWheel = (e: IWheelEvent) => {
-            const delta = this.zoomWheel(e);
+        const zoomFnScrollWheel = (e: IWheelEvent | WheelEvent) => {
+            let ev: any = e;
```
Comment (RaananW) [inline guiEditor/src/diagram/workbench.tsx:858]: Just wondering about typing - are both the interface and the implementation needed? any why is the any casting required here?

**Review by RaananW — COMMENTED** [review]
LGTM! Just a question about the type usage throughout the PR. there is are a few instances of "any", and I wonder if it is needed, especially since there is an interface the supports the values that are "any"ed (IUIEvent?)

**Review by carolhmj — APPROVED** [review]

File: guiEditor/src/diagram/workbench.tsx (hunk 1 above)
Comment (PolygonalSun) [reply guiEditor/src/diagram/workbench.tsx:858]: I'm not sure if this was the best approach but since this can both be called with WheelEvents and IWheelEvents (invoking from our side), I wanted to cover our bases and basically funnel either into an IWheelEvent.  I used `any` here because our custom events now need a deviceType, deviceSlot, and inputIndex value and WheelEvents are readonly.

**Comment by bghgary** [comment]
Reviewed offline. Converting to draft until changes are made.

File: guiEditor/src/diagram/workbench.tsx (hunk 1 above)
Comment (PolygonalSun) [reply guiEditor/src/diagram/workbench.tsx]: reverted this

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts [hunk 2]
```diff
@@ -36,7 +36,7 @@ export class WebDeviceInputSystem implements IDeviceInputSystem {
     }
 
     public onDeviceDisconnected: (deviceType: DeviceType, deviceSlot: number) => void;
-    public onInputChanged: (deviceEvent: IDeviceEvent) => void;
+    public onInputChanged: (eventData: any) => void;
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/webDeviceInputSystem.ts:39]: Why `any` here?

File: src/Inputs/scene.inputManager.ts [hunk 3]
```diff
@@ -267,7 +267,10 @@ export class InputManager {
      * @param pointerEventInit pointer event state to be used when simulating the pointer event (eg. pointer id for multitouch)
      */
     public simulatePointerMove(pickResult: PickingInfo, pointerEventInit?: PointerEventInit): void {
-        let evt = new PointerEvent("pointermove", pointerEventInit);
+        let evt: any = new PointerEvent("pointermove", pointerEventInit);
```
Comment (bghgary) [inline src/Inputs/scene.inputManager.ts:270]: why `any`? same below

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts (hunk 2 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:39]: This isn't supposed to be any; changed to IUIEvent

File: src/Inputs/scene.inputManager.ts (hunk 3 above)
Comment (PolygonalSun) [reply src/Inputs/scene.inputManager.ts]: Removed any, added deviceType, deviceSlot, and inputIndex to UIEvent declaration in browser.d.ts

**Review by bghgary — CHANGES_REQUESTED** [review]
Should not modify browser.d.ts

**Comment by bghgary** [comment]
> Should not modify browser.d.ts

Disregard. Chatted offline.

File: src/LibDeclarations/browser.d.ts [hunk 4]
```diff
@@ -36,6 +36,13 @@ interface CanvasRenderingContext2D {
     msImageSmoothingEnabled: boolean;
 }
 
+// Babylon Extension to enable Native compatibility
```
Comment (bghgary) [inline src/LibDeclarations/browser.d.ts:39]: nit: these properties aren't just for native

**Review by bghgary — APPROVED** [review]

**Comment by bghgary** [comment]
Remember to add to the breaking change section what has changed with DSM contract.

File: src/LibDeclarations/browser.d.ts (hunk 4 above)
Comment (PolygonalSun) [reply src/LibDeclarations/browser.d.ts]: Updated comment to reflect event compatibility, rather than just native compatibility

**Review by RaananW — APPROVED** [review]

**Review by bghgary — CHANGES_REQUESTED** [review]
Please increment the PROTOCOL_VERSION in nativeEngine.ts

**Review by bghgary — COMMENTED** [review]
looks pretty good

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts [hunk 5]
```diff
@@ -403,22 +403,18 @@ export class WebDeviceInputSystem implements IDeviceInputSystem {
...
+                this.onInputChanged(deviceType, deviceSlot, deviceEvent);
 
                 // Lets Propagate the event for move with same position.
                 if (!this._usingSafari && evt.button !== -1) {
-                    deviceEvent.inputIndex = evt.button + 2;
+                    const inputIndex = evt.button + 2;
                     pointer[evt.button + 2] = (pointer[evt.button + 2] ? 0 : 1); // Reverse state of button if evt.button has value
-
-                    this.onInputChanged(deviceEvent);
+                    const ev: IUIEvent = DeviceEventFactory.CreateDeviceEvent(deviceType, deviceSlot, inputIndex, pointer[evt.button + 2], this, this._elementToAttachTo);
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:416]: That was the idea with this change.  I wanted to remove some event ambiguity and ensure that it was a down or up.  Unfortunately, this didn't work for one of my edge case tests so I had to revert it to the previous way of doing it.

File: src/DeviceInput/InputDevices/deviceSourceManager.ts [hunk 6]
```diff
@@ -42,32 +40,56 @@ export class DeviceSourceManager implements IDisposable, IObservableManager {
...
      * @returns All available DeviceSources
      */
     public getDevices(): ReadonlyArray<DeviceSource<DeviceType>> {
-        return this._getDevices();
+        const deviceArray = new Array<DeviceSource<DeviceType>>();
+        this._devices.forEach((deviceSet) => {
+            deviceArray.push.apply(deviceArray, deviceSet);
+        });
+
+        return deviceArray;
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/deviceSourceManager.ts:76]: Changed

File: src/DeviceInput/InputDevices/internalDeviceSourceManager.ts [hunk 7]
```diff
@@ -22,15 +21,20 @@ declare module "../../Engines/engine" {
 /** @hidden */
 export interface IObservableManager {
     onDeviceConnectedObservable: Observable<DeviceSource<DeviceType>>;
-    onInputChangedObservable: Observable<IUIEvent>;
     onDeviceDisconnectedObservable: Observable<DeviceSource<DeviceType>>;
+
+    // Functions
+    _onInputChanged (deviceType: DeviceType, deviceSlot: number, eventData: IUIEvent): void;
+    _addDevice(deviceSource: DeviceSource<DeviceType>): void;
+    _removeDevice(deviceType: DeviceType, deviceSlot: number): void;
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/internalDeviceSourceManager.ts:29]: space removed

File: src/DeviceInput/Helpers/eventFactory.ts [hunk 8]
```diff
@@ -106,20 +106,24 @@ export class DeviceEventFactory {
...
+        // Handle offsets/deltas based on existence of HTMLElement
+        if (elementToAttachTo) {
+            evt.movementX = 0;
+            evt.movementY = 0;
+            evt.offsetX = evt.movementX - elementToAttachTo.getBoundingClientRect().x;
+            evt.offsetY = evt.movementY - elementToAttachTo.getBoundingClientRect().y;
+        }
+        else {
+            evt.movementX = deviceInputSystem.pollInput(deviceType, deviceSlot, 10); // DeltaHorizontal
+            evt.movementY = deviceInputSystem.pollInput(deviceType, deviceSlot, 11); // DeltaVertical
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/DeviceInput/Helpers/eventFactory.ts:119]: Good call, changed

File: src/DeviceInput/InputDevices/nativeDeviceInputSystem.ts [hunk 9]
```diff
@@ -22,18 +51,11 @@ export class NativeDeviceInputSystem implements IDeviceInputSystem {
...
-            const idx = (inputIndex === PointerInput.Horizontal || inputIndex === PointerInput.Vertical || inputIndex === PointerInput.DeltaHorizontal || inputIndex === PointerInput.DeltaVertical) ? PointerInput.Move : inputIndex;
-            const evt = DeviceEventFactory.CreateDeviceEvent(deviceType, deviceSlot, inputIndex, currentState, this);
-
-            let deviceEvent = evt as IDeviceEvent;
-            deviceEvent.deviceType = deviceType;
-            deviceEvent.deviceSlot = deviceSlot;
-            deviceEvent.inputIndex = idx;
-            deviceEvent.previousState = previousState;
-            deviceEvent.currentState = currentState;
+        this._nativeInput.onInputChanged = (deviceType, deviceSlot, inputIndex, previousState, currentState) => {
```
Comment (ryantrem) [inline src/DeviceInput/InputDevices/nativeDeviceInputSystem.ts:54]: I guess `previousState` is no longer used? Should it be removed from here and from BN?

File: src/DeviceInput/InputDevices/deviceEnums.ts [hunk 10]
```diff
@@ -45,11 +45,7 @@ export enum PointerInput {
     MouseWheelY = 8,
     /** Mouse Wheel Z */
     MouseWheelZ = 9,
-    /** Delta X */
-    DeltaHorizontal = 10,
-    /** Delta Y */
-    DeltaVertical = 11,
-    /** Move Catch-all */
+    /** Move Catch-all (Only used in Observable) */
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/deviceEnums.ts:48]: Modified comments for more clarity

File: src/DeviceInput/InputDevices/deviceSource.ts [hunk 11]
```diff
@@ -1,9 +1,15 @@
-import { DeviceType } from './deviceEnums';
+import { DeviceType, PointerInput } from './deviceEnums';
 import { Observable } from '../../Misc/observable';
 import { DeviceInput } from './deviceTypes';
 import { IDeviceInputSystem } from './inputInterfaces';
 import { IUIEvent } from '../../Events/deviceInputEvents';
 
+/** @hidden */
+export type DeviceEventInput<T extends DeviceType> =
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/deviceSource.ts:8]: Nope, it does not.  Removed export

File: src/DeviceInput/InputDevices/deviceSourceManager.ts [hunk 12]
```diff
@@ -17,22 +17,20 @@ export class DeviceSourceManager implements IDisposable, IObservableManager {
...
-     */
-    public readonly onInputChangedObservable: Observable<IUIEvent>;
+
     /**
      * Observable to be triggered when after a device is disconnected
      */
     public readonly onDeviceDisconnectedObservable: Observable<DeviceSource<DeviceType>>;
 
+    /** @hidden */
+    public readonly _onInputChanged: (deviceType: DeviceType, deviceSlot: number, eventData: IUIEvent) => void;
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/deviceSourceManager.ts:27]: Changed to regular function

File: src/DeviceInput/InputDevices/nativeDeviceInputSystem.ts (hunk 9 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/nativeDeviceInputSystem.ts:54]: That would make sense.  Removed from JS.  I'll update my BN PR to reflect this.

File: src/DeviceInput/InputDevices/deviceSource.ts [hunk 13]
```diff
@@ -5,7 +5,7 @@ import { IDeviceInputSystem } from './inputInterfaces';
 import { IUIEvent } from '../../Events/deviceInputEvents';
 
 /** @hidden */
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/deviceSource.ts:7]: nit: this is useless if it's not exported.

**Review by bghgary — APPROVED** [review]
one tiny nit

File: src/DeviceInput/InputDevices/deviceSource.ts (hunk 13 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/deviceSource.ts]: Removed hidden
